src/test/liquidity_tests.cpp

```diff
@@ -0,0 +1,377 @@
+
+    // create pools
+    for (int i = 0; i < PoolCount; ++i) {
+        DCT_ID idA, idB, idPool;
+        std::tie(idA, idB, idPool) = CreatePoolNTokens(mnview, "A"+std::to_string(i), "B"+std::to_string(i));
+        auto optPool = mnview.GetPoolPair(idPool);
+        BOOST_REQUIRE(optPool);
+
+    }
+    // create shares
+    mnview.ForEachPoolPair([&] (DCT_ID const & idPool, CPoolPair const & pool) {
+//            printf("pool id = %s\n", idPool.ToString().c_str());
+        for (int i = 0; i < ProvidersCount; ++i) {
+            CScript shareAddress = CScript(idPool.v * ProvidersCount + i);
+            Res res = AddPoolLiquidity(mnview, idPool, idPool.v*COIN, idPool.v*COIN, shareAddress);
+            BOOST_CHECK(res.ok);
+        }
+        return true;
+    });
+
+    {
+        CCustomCSView cache(mnview);
+
+        // set pool rewards rates
+        const DCT_ID RWD50 = DCT_ID{1};
+        const DCT_ID RWD25 = DCT_ID{2};
+        SetPoolRewardPct(cache, RWD50, COIN/2); // 50%
+        SetPoolRewardPct(cache, RWD25, COIN/4); // 25%
+
+        SetPoolRewardPct(cache, DCT_ID{3}, COIN/10); // 10%
+        SetPoolRewardPct(cache, DCT_ID{4}, COIN/10); // 10%
+        SetPoolRewardPct(cache, DCT_ID{5}, COIN/100);
+        SetPoolRewardPct(cache, DCT_ID{6}, COIN/100);
+        SetPoolRewardPct(cache, DCT_ID{7}, COIN/100);
+        SetPoolRewardPct(cache, DCT_ID{8}, COIN/100);
+        SetPoolRewardPct(cache, DCT_ID{9}, COIN/100);
+        /// DCT_ID{10} - 0
+
+        // set "traded fees" here too, just to estimate proc.load
+        cache.ForEachPoolPair([&] (DCT_ID const & idPool, CPoolPair const & pool) {
+            SetPoolTradeFees(cache, idPool, idPool.v * COIN, idPool.v * COIN*2);
+            return true;
+        });
+
+        // distribute 100 coins
+        CAmount totalRwd = 100*COIN;
+
+        int64_t nTimeBegin = GetTimeMicros();
+        CAmount distributed = cache.DistributeRewards(totalRwd,
+            [&cache] (CScript const & owner, DCT_ID tokenID) {
+                return cache.GetBalance(owner, tokenID);
+            },
+            [&cache] (CScript const & to, CTokenAmount amount) {
+                return cache.AddBalance(to, amount);
+            }
+        );
+        int64_t nTimeEnd = GetTimeMicros(); auto nTimeRwd = nTimeEnd - nTimeBegin;
+        printf("Rewarded %d pools with %d shares each: %.2fms \n", PoolCount, ProvidersCount, 0.001 * (nTimeRwd));
+
+        printf("Distributed: = %ld\n", distributed);
+        BOOST_CHECK(distributed == 9999000000); // always slightly less due to MINIMUM_LIQUIDITY & rounding
+
+        // check it
+        auto rwd25 = 25 * COIN / ProvidersCount;
+        auto rwd50 = 50 * COIN / ProvidersCount;
+        cache.ForEachPoolShare([&] (DCT_ID const & id, CScript const & owner) {
+//            if (id == RWD25/* || id == RWD50*/)
+//                printf("owner = %s: %s,\n", owner.GetHex().c_str(), cache.GetBalance(owner, DCT_ID{0}).ToString().c_str());
```

Comment:
Should we remove debugging code?

Reply:
Sure.. If you think it will not be needed anymore ^)

Reply:
@DeFiDev 
> Sure.. If you think it will not be needed anymore ^)

Debug code should not be committed, esp. for stable version.